### ISSUE TITLE
Compatibility checking should traverse all dependencies

### ISSUE THREAD

**Issue by politas** [KSP-CKAN/CKAN#2231]
Currently, when we do a check for mod availability, we only check one level, and then leave it to the installation process to fail if there are second or third order dependencies that fail. I think our availability checking process should follow this process, and it should be in the core, utilised by all UIs:

![availability flowchart](https://user-images.githubusercontent.com/2109967/34317279-2ac897e0-e7ff-11e7-8a58-70b6d88e15fe.png)

The list of Checked ModVersions should be initialised before starting into this cross-recursive cycle.
I'd appreciate feedback on this. Obviously, this process simply checks availability, without adding any mods to an install list, which would need to be a separate process.

**Comment by politas** [comment]
On review, I think we could gain some efficiencies by maintaining two sets of ModVers - passed and failed, and testing for both of those before calling IsAvailable (ModVersion) in the left-hand flow. 

For the GUIModList, running through this for the whole list could take a long time, and having separate lists of available and not available ModVers built up would certainly reduce that time. Could we do that in the background and update the modlist as we go? 

Also, I can conceive of a false positive result where running on an old KSP version, TopMod depends on Mod1 and Mod2, where Mod1V1 meets compatibility but not Mod1V2, and Mod2 has a dependency on Mod1V2. Not a terribly likely occurrence, but certainly possible.

**Comment by HebaruSan** [comment]
I agree with the title and first paragraph, but I'm "indeterminate" on the details so far.

I feel like there must be a "dynamic programming" approach that we could execute almost for free as we load the registry, but attempting to spell it out has not yet borne fruit. The mods and their dependencies form a giant graph, and we gradually paint more and more parts of it as Incompatible; at the end anything we haven't painted is determined to be Compatible.

**Comment by politas** [comment]
I fear my programming style tends towards "hopelessly procedural". I don't really understand how the core CKAN Registry code works. I cannot see how available_modules is populated.

**Comment by HebaruSan** [comment]
I believe it happens here, along with the rest of the Registry:

https://github.com/KSP-CKAN/CKAN/blob/91e248cc3e1a62241ccc608e17efa625d66d3e15/Core/Registry/RegistryManager.cs#L286-L287

... so if we wanted to run code _literally_ "as we load", it would have to interact with the JSON deserializer to add callbacks (I'm assuming this is possible, I haven't checked for sure). But it would be about the same to run it later; probably should even be on-demand for cases where we don't need dependency info, like `ckan repo list`.

**Comment by HebaruSan** [comment]
... and for completeness, when updating the registry from the network, it happens here (one mod at a time):

https://github.com/KSP-CKAN/CKAN/blob/91e248cc3e1a62241ccc608e17efa625d66d3e15/Core/Net/Repo.cs#L477-L482

**Comment by HebaruSan** [comment]
Possible duplicate of #1646.

**Comment by politas** [comment]
Ok, updated version with some fixes.
![availability flowchart v2](https://user-images.githubusercontent.com/2109967/34480767-5dddf1a6-f002-11e7-9d03-6bca51e8bc1d.png)
I would expect the VersionRange part could be done as an overloaded method, and if there's a specific dependency version, we could just iterate to isAvailable (ModVersion)

Thinking more on the idea of painting incompatibilities as we load the registry, we'd mark a lot of mods as incompatible simply because their dependencies haven't been loaded yet. We _can_ do it as we fill in the MainModList full_list_of_mods, since we already have the whole registry at that point.

**Comment by HebaruSan** [comment]
Yes, unloaded dependencies are a factor. I think the dynamic programming approach would involve an "indeterminate" bucket grouped by dependencies; then after we determine that a given module is incompatible, we would check the indeterminate bucket for mods that depend on it and file them as incompatible as well, and so on up the tree. Still not sure if that really buys us much though.

**Comment by HebaruSan** [comment]
Another thing to consider is how we handle circular dependencies. For example, many of the mods broken out into "core" and "config" components have mutual dependencies back and forth to ensure the core isn't installed without at least one config and vice versa. I think the flow chart if interpreted literally would go into an infinite recursion dealing with that.

**Comment by politas** [comment]
>I think the flow chart if interpreted literally would go into an infinite recursion dealing with that.

Ah, yes! I knew there was a reason I was adding mods to the "checked" list before diving into the dependencies

**Comment by HebaruSan** [comment]
Do you know of any mods affected by this on KSP 1.3.1, for testing purposes? As far as I can tell, this is the complete list:

```
* SSRSS-Clouds-HR (v0.0.3.1) - Stock Size Real Solar System - High Res Clouds
* SSRSS-Clouds-LR (v0.0.3.1) - Stock Size Real Solar System - Low Res Clouds
* SSRSS-Rescale (v0.9.6) - Stock Size Real Solar System - Rescale
* SSRSS-StockIncl (v0.0.3.1) - Stock Size Real Solar System - Stock Inclinations
```

Their grandparent `RealSolarSystem` is 1.2.2-only. All other incompatible mods seem to be either incompatible with KSP 1.3.1 directly, or directly dependent on a mod that is. If you know of others, please cough them up, because it means I have bugs to fix.

**Comment by HebaruSan** [comment]
This is surprisingly difficult to get right. My code struggles with `provides`; there are 2 or 3 1.3.1-compatible mods in the database that provide a name that is itself a real mod that is 1.3.1-incompatible, which messes with my approach to optimization. Between that and the circular dependencies, it feels like the data structure is inherently non-deterministic.

Maybe the status quo is good enough, if SSRSS is the only casualty.
